File: 76.minimum_window_substring/minWindow.java
public class Solution {
    int initTargetHash(int []targethash, String Target) {
        int targetnum =0 ;
        for (char ch : Target.toCharArray()) {
            targetnum++;
            targethash[ch]++;
        }
        return targetnum;
    }
    public String minWindow(String Source, String Target) {
        // queueing the position that matches the char in T
        int ans = Integer.MAX_VALUE;
        String minStr = "";

        int[] targethash = new int[256];

        int targetnum = initTargetHash(targethash, Target);
        int sourcenum = 0;
        int j = 0, i = 0;
        for(i = 0; i < Source.length(); i++) {
            if(targethash[Source.charAt(i)] > 0) //contains the char in target
                sourcenum++;

            targethash[Source.charAt(i)]--;
            while(sourcenum>=targetnum) { //contains all the characters
                if(ans > i - j + 1) { //found a smaller window
                    ans = Math.min(ans, i - j + 1);
                    minStr = Source.substring(j, i + 1);
                }
                targethash[Source.charAt(j)]++;
                if(targethash[Source.charAt(j)] > 0)
                    sourcenum--;
                j ++;
            }
        }
        return minStr;
    }

}

File: 76.minimum_window_substring/minWindow.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    //build the hashtable to count the target char occurrance
    var targethash = [];
    var targetsum = 0;
    for(var m = 0; m < t.length; m++) {
        if (targethash[t.charAt(m)] === undefined) {
            targethash[t.charAt(m)] = 1;
        }else{
            targethash[t.charAt(m)]++;
        }
        targetsum++;
    }

    var i = 0;
    var sourcesum = 0;
    var res = Number.MAX_SAFE_INTEGER;
    var minStr = "";
    for(var j = 0; j < s.length; j++) {
        if (targethash[s.charAt(j)] > 0) {
            sourcesum++;
        }
        targethash[s.charAt(j)]--;

        while (sourcesum >= targetsum) {
            if (res >= j - i + 1) {
                res = Math.min(res, j - i + 1);
                minStr = s.substring(i,j+1);
            }
            targethash[s.charAt(i)]++;//recover the char at i
            if (targethash[s.charAt(i)] > 0) {
                sourcesum--;
            }
            i++;
        }
    }
    return minStr;
};

console.log(minWindow("ADOBECODEBANC","ABC"));